DataFormats/Common/interface/DeviceProduct.h

```diff
@@ -45,7 +47,7 @@ namespace edm {
   template <typename T>
   class DeviceProduct : public DeviceProductBase {
   public:
-    DeviceProduct() = default;
+    explicit DeviceProduct(edm::AllocateForOverwrite) : data_{edm::allocateForOverwrite} {}
```

Comment:
> (just to note, if we'd really want to, we could allow T to not have neither default or AllocateForOverwrite constructor)

how would that work ?
don't we need `edm::Wrapper<T>` to be constructible either by default or passing `kUninitialized` ?

Reply:
OK, I've changed `DeviceProduct` to support products that have either a default constructor, o a constructor that take a `kUninitialized`.

I've tested it reverting locally some of the changes to the device dataformats to use the default constructor, and it seems to work.

I thought about adding some unit tests, but I realised that I have no idea what we should test for :-/

Reply:
> > (just to note, if we'd really want to, we could allow T to not have neither default or AllocateForOverwrite constructor)
> 
> how would that work ? don't we need `edm::Wrapper<T>` to be constructible either by default or passing `kUninitialized` ?

_Theoretically_ we could be able use `std::optional<T>` for `data_` member instead of `T` when `T` does not have either default or `edm::Uninitialized` constructor. But maybe it's not worth of the hassle.

Reply:
Ah, I see... but if we do that couldn't we drop the `edm::Uninitialized` part completely ?

Reply:
> Ah, I see... but if we do that couldn't we drop the `edm::Uninitialized` part completely ?

You're right, I'd think in that case the `edm::Uninitialized` constructor could be omitted. I'd consider also making the default constructor of `DeviceProduct` (or probably the `DeviceProductBase`) to throw an exception (or `assert()`) because it should never be called in practice.